crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3347,19 +3385,58 @@ impl<'db> TypeInferenceBuilder<'db> {
     }
 
     fn infer_slice_expression(&mut self, slice: &ast::ExprSlice) -> Type<'db> {
+        enum SliceArg {
+            Arg(Option<i32>),
+            Unsupported,
+        }
+
         let ast::ExprSlice {
             range: _,
             lower,
             upper,
             step,
         } = slice;
 
-        self.infer_optional_expression(lower.as_deref());
-        self.infer_optional_expression(upper.as_deref());
-        self.infer_optional_expression(step.as_deref());
+        let ty_lower = self.infer_optional_expression(lower.as_deref());
+        let ty_upper = self.infer_optional_expression(upper.as_deref());
+        let ty_step = self.infer_optional_expression(step.as_deref());
 
-        // TODO slice
-        Type::Todo
+        let type_to_slice_argument = |ty: Option<Type<'db>>| match ty {
+            Some(Type::IntLiteral(n)) if i32::try_from(n).is_ok() => {
+                SliceArg::Arg(Some(i32::try_from(n).expect("checked in branch arm")))
+            }
+            Some(Type::BooleanLiteral(b)) => SliceArg::Arg(Some(i32::from(b))),
+            Some(Type::None) => SliceArg::Arg(None),
+            Some(Type::Instance(class)) if class.is_known(self.db, KnownClass::NoneType) => {
+                SliceArg::Arg(None)
+            }
+            None => SliceArg::Arg(None),
+            _ => SliceArg::Unsupported,
+        };
+
+        match (
+            type_to_slice_argument(ty_lower),
+            type_to_slice_argument(ty_upper),
+            type_to_slice_argument(ty_step),
+        ) {
+            (SliceArg::Arg(lower), SliceArg::Arg(upper), SliceArg::Arg(step)) => {
+                match step.map(NonZeroI32::new) {
+                    Some(Some(step)) => {
+                        Type::SliceLiteral(SliceLiteralType::new(self.db, lower, upper, Some(step)))
+                    }
+                    None => Type::SliceLiteral(SliceLiteralType::new(self.db, lower, upper, None)),
+                    Some(None) => {
+                        self.add_diagnostic(
+                            AnyNodeRef::ExprSlice(slice), // TODO
+                            "zero-stepsize-in-slice",
+                            format_args!("Slice step size can not be zero"),
+                        );
```

Comment:
In a previous version of this PR, I deferred reporting this error to the usage site where we try to slice a string/bytes-literal or a tuple. Raising the diagnostic here is too early. The slice might be used on a type with a custom `__getitem__` impl that makes use of `0` as a step size somehow.

I'll fix this.